File: elevator-system/java-without-multithreading/ElevatorController.java
import java.util.ArrayList;
import java.util.List;

/**
 * Manages a fleet of elevators, distributing requests optimally.
 */
public class ElevatorController {
    private final List<Elevator> elevators;

    public ElevatorController(int numElevators, int capacity) {
        elevators = new ArrayList<>();
        for (int i = 0; i < numElevators; i++) {
            elevators.add(new Elevator("E" + (i + 1), capacity));
        }
    }

    /**
     * Finds the closest suitable elevator for a request and assigns it.
     */
    public void requestElevator(Request req) throws Exception {
        Elevator optimalElevator = null;
        int minDistance = Integer.MAX_VALUE;

        for (Elevator el : elevators) {
            int dist = el.getDistanceIfAssigned(req);
            if (dist < minDistance) {
                minDistance = dist;
                optimalElevator = el;
            }
        }

        if (optimalElevator == null) {
            throw new Exception("all elevators are currently over capacity to handle " + req.getPassengers() + " passengers");
        }

        System.out.printf("[Dispatcher] \uD83D\uDFE2 Routing (Src: %d -> Dest: %d, Pax: %d) into Optimal => %s\n",
                req.getSourceFloor(), req.getDestinationFloor(), req.getPassengers(), optimalElevator.getId());
        
        optimalElevator.addRequest(req);
    }

    /**
     * Advances the simulation by one step for all elevators.
     * @return true if any elevator is currently active.
     */
    public boolean tickAll() {
        boolean isActive = false;
        for (Elevator el : elevators) {
            el.tick();
            if (el.getCurrentDirection() != Direction.IDLE) {
                isActive = true;
            }
        }
        return isActive;
    }
}

File: elevatorSystem/go-without-multithreading/elevator_controller.go
package main

import (
	"fmt"
)

// ElevatorController manages a fleet of elevators and dispatches incoming requests
type ElevatorController struct {
	Elevators []*Elevator // array of all managed elevators
}

// NewElevatorController initializes the controller with a specified number of elevators and capacity
func NewElevatorController(numElevators int, capacity int) *ElevatorController {
	elevators := make([]*Elevator, numElevators)
	for i := 0; i < numElevators; i++ {
		elevators[i] = NewElevator(fmt.Sprintf("E%d", i+1), capacity)
	}
	return &ElevatorController{
		Elevators: elevators,
	}
}

// RequestElevator assigns a passenger request to the most optimal elevator based on distance and capacity
func (ec *ElevatorController) RequestElevator(req *Request) error {
	var optimalElevator *Elevator
	minDistance := int(1e9)

	// Loop through all elevators to find the closest one capable of handling the request
	for _, el := range ec.Elevators {
		dist := el.GetDistanceIfAssigned(req)
		if dist < minDistance {
			minDistance = dist
			optimalElevator = el
		}
	}

	if optimalElevator == nil {
		return fmt.Errorf("all elevators are currently over capacity to handle %d passengers", req.Passengers)
	}

	fmt.Printf("[Dispatcher] 🟢 Routing (Src: %d -> Dest: %d, Pax: %d) into Optimal => %s\n", req.SourceFloor, req.DestinationFloor, req.Passengers, optimalElevator.ID)
	return optimalElevator.AddRequest(req) // Assign the request to the chosen elevator
}

// TickAll advances the physical simulation by one unit of time for all managed elevators
// Returns true if at least one elevator is still active (moving or processing stops)
func (ec *ElevatorController) TickAll() bool {
	isActive := false
	for _, el := range ec.Elevators {
		el.Tick()
		if el.CurrentDirection != IDLE {
			isActive = true
		}
	}
	return isActive
}
